### ISSUE TITLE
How mutable should PKEYs be after initialisation?

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#14075]
The possibility was raised first here: https://github.com/openssl/openssl/pull/14062#issuecomment-773112106 :

> I see another potential race condition:
> 
>     thread 1 has an ongoing operation (digestverify, say)
>     thread 2 makes a change to the key (calls EVP_PKEY_set1_encoded_public_key(), say) and starts another operation (digestsign)
> 
> Thread 2 would invalidate the operation cache, because the dirty count doesn't match up, which would free all the keydata, including the one that thread 1 is currently using.
> 
> Whoops.

Some discussion follows from: https://github.com/openssl/openssl/pull/14062#issuecomment-773225594

**Comment by beldmit** [comment]
For me it looks like the question is not about mutability but about thread-safety

**Comment by kroeckx** [comment]
There is also discussion about that in https://github.com/openssl/openssl/pull/14059#discussion_r569793419

**Comment by levitte** [comment]
> For me it looks like the question is not about mutability but about thread-safety

It's really about both.  Like I've said elsewhere (I forget where), there was a discussion years ago (long before 2014) around the idea of declaring a key "finished".  That would be a state where a key wouldn't mutate any more, and could be used for sign, verify, etc operations.  And yes, that *is* a way of fixing the threading problem, but it's also a conceptual shift of what you're permitted to do with a key at diverse moments of its life span.

**Comment by paulidale** [comment]
This would go back to the state machine logic restricting which calls can be made on the various EVP contexts.

**Comment by t8m** [comment]
My proposal is:
1. Keys are mutable only up to a point where they are set into an operation context (this would be new in 3.0 but would really simplify things). We do not need to necessarily enforce this at all the possible places but major calls should be blocked and error returned instead of doing something undefined.
2. Mutating keys from multiple threads at once is undefined operation (this is already in 1.1.1).
3. Using keys for different operations (read-only) from multiple threads at once must be possible (this worked in 1.1.1, should keep working).

**Comment by beldmit** [comment]
I agree with 1 and 2, but not sure which operations are really read-only.

**Comment by levitte** [comment]
Number one is fairly easy.  If there's anything in the operation cache, no more modifications are allowed.  The check for this should be in `EVP_PKEY_set_params()` (which is becoming the universal function called from everything else that wants to set params, as off PR #14056, which I just noticed has been merged).

**Comment by kroeckx** [comment]
It seems that the only reason why keys now need to be mutable is
to do a key derivation. We should probably consider a new API
where it's easy to avoid that, but that can wait until after 3.0.

**Comment by levitte** [comment]
> I agree with 1 and 2, but not sure which operations are really read-only.

That's fairly well define by 1, isn't it?  When any operation starts using a key, it becomes unmutable.
(do we want it to be unmutable for the rest of its life span, or is it mutable again when no operation is using it?  Is that the question you're asking, @beldmit?)

**Comment by t8m** [comment]
> I agree with 1 and 2, but not sure which operations are really read-only.

All operations except setting key parameters and a key generation (including generation of a public key from a private key) are read-only.

**Comment by beldmit** [comment]
Not sure. Implementations can do some caching on-demand.

**Comment by t8m** [comment]
> Not sure. Implementations can do some caching on-demand.

If you mean to do some precomputations and cache the values for further operations that could find those values useful, then this is a purely provider side thing and the provider should be fully responsible for properly locking the cache accesses. It is not really a mutability of the key as seen from the application side.

**Comment by levitte** [comment]
What sort of caching are we talking about here?  Of course, providers can do whatever they want with their own data, but it sounds absolutely unwise to cache something operation related into the keydata itself...

**Comment by beldmit** [comment]
Ok, agreed.

**Comment by richsalz** [comment]
A clarification question.  Is pre-computing montgomery or CRT values considered to be the caching in an RSA key? When is that done, and does it properly discard and recompute when the key is modified?  Because to an application signing is a read-only use.

**Comment by levitte** [comment]
> A clarification question. Is pre-computing montgomery or CRT values considered to be the caching in an RSA key?

Isn't that done early, when the key is created, i.e. before any operation uses the key?

**Comment by richsalz** [comment]
> Isn't that done early, when the key is created, i.e. before any operation uses the key?

I don't know, which is why I asked.  I looked at ossl_rsa.c and got too deep into code I don't know.

**Comment by romen** [comment]
I haven't checked closely again in some time, but I am fairly sure that both RSA and EC tended to create internal context to accelerate some operations only on-demand if that operation is called or if the "owner" of the key object called a precomputation function explicitly (probably to avoid precomputing stuff that consumes time and space for every instantiated key object, as maybe that key object is just being duplicated and passed around for different protocol contexts or to convert it to other forms and shapes).
This precomputation usually has its full potential only if it is reused many times for the same key, so it made (and makes) no sense storing it into the short lived operation ctx, which explains why it is altering the contents of the key object.

Now that I think more about it, during the last year I encountered at least one place that had this behavior while I was trying to constify key parameters in EVP or provider functions, but I am failing at finding it!

**Comment by paulidale** [comment]
There is additional discussion in #14391

**Comment by mattcaswell** [comment]
OTC: We will not make EVP_PKEYs fully immutable for 3.0

**Comment by t8m** [comment]
Changing EVP_PKEYs to be fully immutable would be a quite big API break now. Moving to 4.0.0 milestone.

**Comment by t8m** [comment]
We are slowly moving towards the immutability with new algorithms and APIs, etc. There is no point in keeping this issue open though as there are no concrete steps planned.
